### ISSUE TITLE
Investigate using linearized errors for HDR trials

### ISSUE THREAD

**Issue by solidpixel** [ARM-software/astc-encoder#364]
Currently HDR error computation in `compute_symbolic_block_difference()` uses the difference between the LNS-encoded values to determine the quality of an encoding trial, not the difference between the actual linear color values. 

We should investigate computing the linear error to see if it makes a material difference to image quality.

**Comment by solidpixel** [comment]
Initial investigations for this:

- On average PSNR improves significantly (+1-2dB not uncommon). 
- On average mPSNR worsens significantly (-0.5dB not uncommon).

Visually low-bitrate encodings improve quite a lot - manual inspection shows a significant reductions in block artifacts - so this looks like a reasonable change to make. However, a new problem that this introduces is that very bright texels in a block are prone to dominating the error metrics. The compressor will sacrifice a lot of quality in neighbouring texels to improve the accuracy of the bright texels - often to the point that they quantize to black.

**Comment by solidpixel** [comment]
Yes, agreed. At this point I'm mostly just experimenting to see what helps. The current compressor is pretty strong at 8bpp, but suffers horrible block artefacts at lower bitrates in regions with high brightness and complex patterns. It may be the end result is that lower bitrates require content constraints to avoid this - email.g. limiting max brightness much like an RGBM texture.

**Comment by solidpixel** [comment]
This type of block error seems to be a particular quirk of the LNS encoding, rather than the increased dynamic range in the image. 

Even with max brightness clamped to 1.0, the HDR image still suffers from severe block artifacts for 6x6 blocks, whereas the LDR equivalent compresses no problem.

**Comment by TomOlson** [comment]
Cool image, thanks for sharing! I don't remember the details of LNS - it's that piecewise linear approximation to a log curve, right? I have to say, it _looks_ like a fair number of blocks in the image are mapping to a single color, which feels like a bug. Unless the LNS color mode somehow constrains one of the color endpoints to be (0,0,0), or something? I have not read the spec since around 2015 or so...

**Comment by solidpixel** [comment]
> LNS - it's that piecewise linear approximation to a log curve, right?

Yes, that's the one. 

> I have to say, it looks like a fair number of blocks in the image are mapping to a single color, which feels like a bug.

Yeah, something definitely smells with this. I just wish I knew what =) The code is generic and "works" for 8bpp, so if it is a bug it's a subtle one ... 

For the cases where there is actual high dynamic range I can sort of understand the problem from a conceptual point of view. The endpoints are far apart, and so you need very fine weight quantization to get granularity of texel weights along the luminance spectrum. 

I don't understand why LNS it's so bad when brightness is clamped < 1.0. That shouldn't be any worse than the LDR compression in terms of endpoint spacing, so that's the obvious initial line of investigation.

**Comment by solidpixel** [comment]
Some observations from Rich G who has made HDR 6x6 play nicely with basisu:

* https://github.com/richgel999/junkdrawer/wiki/ASTC-6x6-HDR:-astcenc-%E2%80%90thorough-%E2%80%90exhausive-vs.-basis-universal-comp_level-3

**Comment by solidpixel** [comment]
Notes/ideas:
* `compute_quantized_weights_for_decimation()` always chooses between `floor(ideal)` and `ceil(ideal)` based on the midpoint, ignoring the non-linearity of the LNS curve.
